opampserver/pkg/server/handlers.go

```diff
@@ -191,3 +196,68 @@ func (c *ConnectionHandlers) UpdateInstrumentationInstanceStatus(ctx context.Con
 
 	return nil
 }
+
+// resolveK8sAttributes resolves K8s resource attributes using either direct attributes from opamp agent or device cache
+func (c *ConnectionHandlers) resolveK8sAttributes(ctx context.Context, attrs opampAgentAttributesKeys,
+	deviceId string, kubeClient *kubernetes.Clientset, logger logr.Logger) (*di.K8sResourceAttributes, *corev1.Pod, error) {
+
+	if attrs.hasRequiredAttributes() {
+		podInfoResolver := di.NewK8sPodInfoResolver(logger, kubeClient)
+		return resolveFromDirectAttributes(ctx, attrs, podInfoResolver, kubeClient)
+	}
+	return c.deviceIdCache.GetAttributesFromDevice(ctx, deviceId)
+}
+
+func extractOpampAgentAttributes(attributes []*protobufs.KeyValue) opampAgentAttributesKeys {
+	result := opampAgentAttributesKeys{}
+
+	for _, attr := range attributes {
+		switch attr.Key {
+		case string(semconv.TelemetrySDKLanguageKey):
+			result.ProgrammingLanguage = attr.Value.GetStringValue()
+		case string(semconv.ContainerNameKey):
+			result.ContainerName = attr.Value.GetStringValue()
+		case string(semconv.K8SPodNameKey):
+			result.PodName = attr.Value.GetStringValue()
+		case string(semconv.K8SNamespaceNameKey):
+			result.Namespace = attr.Value.GetStringValue()
+		case "k8s.workload.kind":
```

Comment:
This one should be a non-identifying attribute, as it is not a resource attribute